@here/harp-mapview/lib/ClipPlanesEvaluator.ts

```diff
@@ -756,72 +750,72 @@ export class TiltViewClipPlanesEvaluator extends TopViewClipPlanesEvaluator {
+        const d = camera.position.length();
+
         if (camera instanceof THREE.PerspectiveCamera) {
             const halfVerticalFovAngle = THREE.MathUtils.degToRad(camera.fov / 2);
 
-            // Ratio of the depth of the camera compared to the distance between the center of the
-            // screen and the middle of the top of the screen (here set to be value 1). We just need
-            // this to be a ratio because we are interested in just computing the direction of the
-            // upper right corner and not any specific length.
-            const zLength = 1 / Math.tan(halfVerticalFovAngle);
-
-            // Camera space direction vector along the top right frustum edge (of the camera).
-            const upperRightDirection = this.m_tmpVectors[2];
-            upperRightDirection.set(camera.aspect, 1, -zLength);
-
             // Now we need to account for camera tilt and frustum volume, so the longest
             // frustum edge does not intersects with sphere, it takes the worst case
             // scenario regardless of camera tilt, so may be improved little bit with more
             // sophisticated algorithm.
             viewRanges.near *= Math.cos(halfVerticalFovAngle);
-            // Step-wise calculate angle between camera eye vector and tangent
-
-            // Calculate angle between surface normal(below camera position) and tangent.
-            const alpha = Math.asin(r / d);
 
-            // Calculate angle between look at and surface normal(below camera position)
-            const cameraPitch = this.getCameraPitch(cameraToOrigin, camera);
+            // Ratio of the depth of the camera compared to the distance between the center of the
+            // screen and the middle of the top of the screen (here set to be value 1). We just need
+            // this to be a ratio because we are interested in just computing the direction of the
+            // upper right corner and not any specific length.
+            const zLength = 1 / Math.tan(halfVerticalFovAngle);
 
-            // Calculate angle between camera eye vector and tangent (center of the top of the
-            // screen)
-            const modifiedAlpha = Math.abs(alpha - cameraPitch);
+            // Camera space direction vector along the top right frustum edge (of the camera).
+            const upperRightDirection = this.m_tmpVectors[0]
+                .set(camera.aspect, 1, -zLength)
+                .applyMatrix4(camera.matrixWorld)
+                .sub(camera.position);
 
-            // Transform the vector to world space
-            upperRightDirection.applyMatrix4(camera.matrixWorld);
-            // Transform the vector to be relative from the camera
-            upperRightDirection.sub(camera.position);
             // Ray must be given normalized vector, otherwise it won't work
-            upperRightDirection.normalize();
+            this.m_ray.set(camera.position, upperRightDirection.normalize());
```

Comment:
can you bring the `upperRightDirection.normalize()` into the definition?

Reply:
ditto, no need to change since this won't be here in next PR.

Reply:
Ok, cool.